components/net/fetch/methods.rs

```diff
@@ -866,6 +869,405 @@ async fn scheme_fetch(
+        /*
+        Checking for an opaque origin may not fully guarantee that the request is safe
+        (knowingly requested by the local-machine owner) rather than unsafe (came from
+        third-party JavaScript for example) so more work needs to be done to consider how to
+        guarantee safety here.
+        This check should reject requests from pages fetched from the public internet over
+        HTTP[S], FTP, WS[S], but might not stop other public scheme types from generating a
+        local directory listing. (This is why this local directory listing feature is disabled
+        by default, only enabled when the "browser.local_directory_listing.enabled" preference
+        has been set to true.)
+        TODO: Work out whether more work is needed to protect access to local directory listing.
+        */
+        return Response::network_error(NetworkError::Internal(
+            "Cannot request local directory listing from non-local origin.".to_string(),
+        ));
+    }
+    let url = if !url.path().ends_with('/') {
+        // Re-read the path as a directory, so that Servo adds the
+        // forward-slash to the end of the URL (at least internally,
+        // though the URL bar does not reflect this) and loads the
+        // linked directories and files correctly when clicked.
+        if let Ok(dir_url) = Url::from_directory_path(url.path()) {
+            ServoUrl::from_url(dir_url)
+        } else {
+            return Response::network_error(NetworkError::Internal(format!(
+                "Unable to parse local directory path {}",
+                url
+            )));
+        }
+    } else {
+        url
+    };
+    let mut response = Response::new(url, ResourceFetchTiming::new(request.timing_type()));
+    response.headers.typed_insert(ContentType::html());
+    let directory_summary = build_directory_summary(file_path);
+    let page_text = build_html_directory_listing(directory_summary);
+    let bytes: Vec<u8> = page_text.into_bytes();
+    *response.body.lock().unwrap() = ResponseBody::Done(bytes);
+    response
+}
+
+// TODO: WORK OUT WHY pub(crate) DOES NOT WORK HERE - VISIBILITY SHOULD NOT BE FULLY PUBLIC !!!
+// TODO: FIND A WAY TO MAKE THIS VISIBLE ONLY TO THIS FILE AND THE UNIT TESTS !!!
+pub struct DirectorySummary {
+    pub path: Result<String, &'static str>,
+    pub has_parent: bool,
+    pub items: Result<BTreeMap<OsString, DirectoryItemDescriptor>, &'static str>,
+}
+
+// TODO: WORK OUT WHY pub(crate) DOES NOT WORK HERE - VISIBILITY SHOULD NOT BE FULLY PUBLIC !!!
+// TODO: FIND A WAY TO MAKE THIS VISIBLE ONLY TO THIS FILE AND THE UNIT TESTS !!!
+pub fn build_directory_summary(file_path: std::path::PathBuf) -> DirectorySummary {
+    let path = file_path
+        .to_str()
+        .map(str::to_string)
+        .ok_or("Invalid directory path.");
+    let has_parent = file_path.parent().is_some();
+    let items = if let Ok(entries) = std::fs::read_dir(file_path) {
```

Comment:
Interesting idea. I'm trying to work out whether the trait implementations would end up so complicated that they'd need unit tests (which read the actual file system) to confirm their correctness. The structure needs to contain/provide quite a bit of data (file path, parent, metadata, etc), so I don't think a single trait could cover it all.

Reply:
To my surprise, Rust is allowing me to simply `impl Into<DirectorySummary> for PathBuf` so I'm going to use that, allowing unit tests to just mock-up a `DirectorySummary` without any actual reading of real directories. Then get rid of the unit tests which create and read from temporary directories.

That does mean that we'll have to take on faith that the implementation of `Into<DirectorySummary>` works. (We know it does at the moment, because of the real-directory unit tests that were run until now.)